### PR TITLE
TG-809

### PR DESCRIPTION
**Pull request by FernandezAstor**

### CODE REVIEW

File: dk-installer.py [hunk 1]
```diff
@@ -1431,9 +1452,12 @@ class TestGenPullImagesStep(Step):
 
     def execute(self, action, args):
         remaining_attemps = TESTGEN_PULL_RETRIES
+        docker_compose_file_path = get_docker_compose_path()
         while True:
             try:
-                with action.start_cmd("docker", "compose", "pull", "--policy", "always") as (proc, stdout, stderr):
+                with action.start_cmd("docker", "compose", "-f", docker_compose_file_path, "pull") as (
```
Comment (rboni-dk) [inline dk-installer.py:1458]: if I understood the logic correct, this will crash on linux, because `get_docker_compose_path()` would return `None`

File: dk-installer.py [hunk 2]
```diff
@@ -1363,8 +1384,8 @@ def create_compose_file(self, file, username, password, port, image, ssl_cert_fi
                     source: {ssl_key_file}
                     target: /dk/ssl/cert.key 
         """ if ssl_cert_file and ssl_key_file else ""
-        
-        file.write_text(
+        compose_file_path = get_docker_compose_path()
+        compose_file_path.write_text(
```
Comment (rboni-dk) [inline dk-installer.py:1388]: if I understood the logic correct, this will crash on linux, because `get_docker_compose_path()` would return `None`

File: dk-installer.py [hunk 3]
```diff
@@ -1154,19 +1180,20 @@ class DemoContainerAction(Action):
...
-            "--name",
-            DEMO_CONTAINER_NAME,
-            "--network",
-            DOCKER_NETWORK,
-            "--add-host",
-            "host.docker.internal:host-gateway",
-            DEMO_IMAGE,
-            command,
+                "docker",
+                get_docker_compose_path(),
```
Comment (rboni-dk) [inline dk-installer.py:1184]: if I understood the logic correct, this will crash on linux, because `get_docker_compose_path()` would return `None`

Why the extra indentation level?

File: dk-installer.py [hunk 4]
```diff
@@ -1051,19 +1077,19 @@ def execute(self, args):
...
+                    args.profile,
+                    "--",
+                    "--namespace",
+                    args.namespace,
+                    "--address",
+                    "0.0.0.0",
+                    "port-forward",
+                    "service/observability-ui",
+                    f"{args.port}:http",
+                    raise_on_non_zero=False,
```
Comment (rboni-dk) [inline dk-installer.py:1092]: Is this change intentional? The indentation seems off (8 spaces).

I actually see this in a few other places.

File: dk-installer.py [hunk 5]
```diff
@@ -607,7 +631,9 @@ def execute(self, args):
 
 
 class Installer:
-    def __init__(self):
+    def __init__(self, arguments=None):
+        if arguments is not None:
+            sys.argv[1:] = arguments
```
Comment (rboni-dk) [inline dk-installer.py:636]: I think it's a bad idea to update the contents of `sys.argv`. The arg parser has a way to set default arguments. I think this has to be used instead

File: dk-installer.py [hunk 6]
```diff
@@ -496,21 +512,29 @@ def start_cmd(self, *cmd, raise_on_non_zero=True, env=None, **popen_args):
         started = time.time()
         self._cmd_idx += 1
 
-        LOG.debug("Command [%04d]: [%s]", self._cmd_idx, " ".join(cmd))
+        flattened_cmd = [str(item) for sublist in cmd for item in
+                         (sublist if isinstance(sublist, tuple) else [sublist])]
+
+        LOG.debug("Command [%04d]: [%s]", self._cmd_idx, " ".join(flattened_cmd))
```
Comment (rboni-dk) [inline dk-installer.py:518]: I'm not a fan of this logic, especially inside this function that is already doing too much things.

Instead, you can use the star operator (`*`) to expand the list of command pieces inside another list of a function call. That would be much cleaner.

Also, if you go with the approach I'm suggesting later of creating a function to run the docker commands, most of the call sites that are requiring this will vanish.

File: dk-installer.py [hunk 7]
```diff
@@ -335,8 +344,15 @@ class Action:
 
     @contextlib.contextmanager
     def init_session_folder(self, prefix):
-        script_path = pathlib.Path(sys.argv[0]).absolute()
-        data_folder = script_path.parent.joinpath(".dk-installer")
+        if detect_os() == 'Windows':
+            home_dir = pathlib.Path.home()
+            data_folder = home_dir.joinpath("Documents", "dk-installer", "logs")
+            if not data_folder.exists():
+                data_folder.mkdir(parents=True, exist_ok=True)
```
Comment (rboni-dk) [inline dk-installer.py:351]: You don't have to check if the folder exists, because you are calling `mkdir` with `ok_exists=True`. Also, this folder is being created anyway just a few lines below, so you don't even need these two lines

```suggestion
```

File: dk-installer.py [hunk 8]
```diff
@@ -274,6 +282,7 @@ class Requirement:
 
     def check_availability(self, action, args):
         try:
+            self.cmd = get_docker_compose_base_command()
```
Comment (rboni-dk) [inline dk-installer.py:285]: You shouldn't need this change, unless I'm missing something. This will likely break the Observability install. 

```suggestion
```

File: dk-installer.py [hunk 9]
```diff
@@ -125,7 +125,15 @@ def generate_password():
...
-        file_name = CREDENTIALS_FILE.format(product)
+        if detect_os() == 'Windows':
+            home_dir = pathlib.Path.home()
+            credentials_folder = home_dir.joinpath("Documents", "dk-installer")
+            if not credentials_folder.exists():
+                credentials_folder.mkdir(parents=True, exist_ok=True)
+            os.makedirs(credentials_folder, exist_ok=True)
+            file_name = os.path.join(credentials_folder, CREDENTIALS_FILE.format(product))
+        else:
+            file_name = CREDENTIALS_FILE.format(product)
```
Comment (rboni-dk) [inline dk-installer.py:136]: You don't need to check the folder existence, given you are calling `mkdir` with `exist_ok=True`. It's also doing it multiple times. I'm also suggesting some readability/DRY changes.

```suggestion
        if detect_os() == 'Windows':
            credentials_folder = pathlib.Path.home() / "Documents" / "dk-installer"
            credentials_folder.mkdir(parents=True, exist_ok=True)
        else:
            credentials_folder = pathlib.Path()
        file_name =  credentials_folder / CREDENTIALS_FILE.format(product)
```

File: dk-installer.py [hunk 10]
```diff
@@ -1653,23 +1675,29 @@ def _delete_project(self, args):
...
-                CONSOLE.msg("Docker network deleted")
+        start_time = time.time()
+        try:
+            self.run_cmd(
+                "docker",
+                get_docker_compose_path(),
+                "network",
+                "rm",
+                DOCKER_NETWORK,
+                timeout=60,
```
Comment (rboni-dk) [inline dk-installer.py:1691]: When a timeout happens, the `subprocess.TimeoutExpired` exception is raised, and it shouldn't be needed to measure the elapsed time. Also, it's probably not working as you expect exactly because this exception will prevent  code bellow the `run_cmd()` to execute. You will probably need to use the `start_cmd()` instead. There's an example of this approach for pulling the Testgen images.

Also, could you add a constant for the timeout period?

File: dk-installer.py [hunk 11]
```diff
@@ -1791,7 +1818,7 @@ def execute(self, args):
             )
 
         for command in cli_commands:
-            CONSOLE.msg(f"Running command : {' '.join(docker_exec_command)} {' '.join(command)}")
+            CONSOLE.msg(f"Running command : {''.join(str(docker_exec_command))} {''.join(command)}")
```
[the remark this answers is not in the document]
Comment (FernandezAstor) [reply dk-installer.py:1821]: I can't say what I changed but somehow one of my changes added blanks between each command.

This is before my changes:
`| Running command : [ ( ' d o c k e r ' ,   ' c o m p o s e ' ,   ' - f ' ,   ' C : \ \ U s e r s \ \ A s t o r \ \ D o c u m e n t s \ \ d k - i n s t a l l e r \ \ d o c k e r - c o m p o s e . y m l ' ,   ' e x e c ' ,   ' e n g i n e ' ) ] testgen quick-start --delete-target-db`

This is how it looks after my changes:
`| Running command : [('docker', 'compose', '-f', 'C:\\Users\\Astor\\Documents\\dk-installer\\docker-compose.yml', 'exec', 'engine')] testgenquick-start--delete-target-db`

File: dk-installer.py [hunk 12]
```diff
@@ -1803,23 +1830,25 @@ class TestgenDeleteDemoAction(DemoContainerAction):
     def execute(self, args):
         CONSOLE.title("Delete TestGen demo")
         try:
-            self.run_dk_demo_container("tg-delete-demo")
+            if detect_os() == 'Windows':
+                pass
```
[the remark this answers is not in the document]
Comment (rboni-dk) [reply dk-installer.py:1834]: If it's not working, isn't it the case to fix in a separated PR instead of deleting? Not sure I'm following =)

File: dk-installer.py [hunk 13]
```diff
@@ -1854,4 +2015,13 @@ def execute(self, args):
...
     )
-    sys.exit(installer_instance.run())
+    installer_instance.run()
+
+
+if __name__ == "__main__":
+    arguments = sys.argv[1:]
+    print("DataKitchen Installer")
+    if len(arguments) == 0:
+        show_main_menu()
```
[the remark this answers is not in the document]
Comment (aarthy-dk) [reply dk-installer.py:2025]: Let's start with showing the menu only for the windows .exe. We don't want the current installation experience being affected by the changes.

[the remark this answers is not in the document]
Comment (rboni-dk) [reply dk-installer.py:2025]: Then, won't this PR be ever merged? How are we planning to keep this "special version"?

[the remark this answers is not in the document]
Comment (aarthy-dk) [reply dk-installer.py:2025]: I mean shouldn't there be a condition to check whether we are using windows or not? And only if that condition is true, the menu is displayed. Otherwise, the default is how things currently work.

[the remark this answers is not in the document]
Comment (FernandezAstor) [reply dk-installer.py:2025]: I made a function that checks if the code is being run in .py or .exe. If it's .exe the menu will be shown.

File: dk-installer.py [hunk 14]
```diff
@@ -1467,41 +1498,38 @@ class TestGenStartStep(Step):
...
 
 
 class TestGenSetupDatabaseStep(Step):
     label = "Initializing the platform database"
 
     def execute(self, action, args):
         action.run_cmd(
-            "docker",
-            "compose",
+            get_docker_compose_base_command() + (
```
[the remark this answers is not in the document]
Comment (rboni-dk) [reply dk-installer.py:1526]: Alternatively, your function can only deal with building the command parts, which I think it would be even better, especially if you end up using these commands also with the `start_cmd()`.

The call would look like this:

```python
 action.run_cmd(
    *tg_compose_cmd(
        "exec",
        "engine",
        "testgen",
        "setup-system-db",
         "--yes",
    )
)
```

File: dk-installer.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (FernandezAstor) [reply dk-installer.py:1834]: Digging a little more, I found that to delete the demo, self.run_dk_demo_container("tg-delete-demo") runs a command that requires observability to be installed. It's not a Windows issue.

[the remark this answers is not in the document]
Comment (aarthy-dk) [reply dk-installer.py:1834]: If Observability is not installed, `run_dk_demo_container` will fail. But that's why it's wrapped in try-catch-pass. we shouldn't need any additional conditions here.
